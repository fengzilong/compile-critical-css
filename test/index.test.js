const { compile } = require( '../' )
const vm = require( 'vm' )

function buildFn( fnString ) {
  const context = {
    fn: null
  }

  const script = new vm.Script( `fn = ${ fnString }` )
  vm.createContext( context )
  script.runInContext( context )

  return context.fn
}

test( `run compiled function`, () => {
  const fnString = compile( `.a { color: #111; }.b { color: #222; }` )

  const getCritical = buildFn( fnString )

  expect( getCritical( [
    'a'
  ] ) ).toBe( `.a { color: #111; }` )

  expect( getCritical( [
    'b'
  ] ) ).toBe( `.b { color: #222; }` )

  expect( getCritical( [
    'a',
    'b'
  ] ) ).toBe( `.a { color: #111; }.b { color: #222; }` )
} )

test( `preserve media rule`, () => {
  const fnString = compile( `
    @media screen and (min-width: 900px) {
      article {
        padding: 1rem 3rem;
      }
      .a { color: #111; }
      .b { color: #222; }
    }
  ` )

  const getCritical = buildFn( fnString )

  expect( getCritical( [
    'a'
  ] ) ).toMatchSnapshot()
} )

test( `preserve nested at rule`, () => {
  const fnString = compile( `
    @supports (display: flex) {
      @media screen and (min-width: 900px) {
        article {
          display: flex;
        }
        .a { color: #111; }
        .b { color: #222; }
      }
    }
  ` )

  const getCritical = buildFn( fnString )

  expect( getCritical( [
    'a'
  ] ) ).toMatchSnapshot()
} )

test( `preserve import`, () => {
  const fnString = compile( `
    @import url("fineprint.css") print;

    @supports (display: flex) {
      @media screen and (min-width: 900px) {
        article {
          display: flex;
        }
        .a { color: #111; }
        .b { color: #222; }
      }
    }
  ` )

  const getCritical = buildFn( fnString )

  expect( getCritical( [
    'a'
  ] ) ).toMatchSnapshot()
} )

test( `preserve css variable`, () => {
  const fnString = compile( `
    :root {
      --main-bg-color: brown;
    }

    element {
      background-color: var(--main-bg-color);
    }

    .a { color: var(--main-bg-color); }
    .b { color: var(--main-bg-color); }
  ` )

  const getCritical = buildFn( fnString )

  expect( getCritical( [
    'a'
  ] ) ).toMatchSnapshot()
} )

test( `preserve ::pseudo`, () => {
  const fnString = compile( `
    .element {
      background-color: var(--main-bg-color);
    }

    .a::before, .b::before { color: var(--main-bg-color); }
    .a {
      color: #ddd;
    }
    .b::before { color: var(--main-bg-color); }
    .b { color: var(--main-bg-color); }
  ` )

  const getCritical = buildFn( fnString )

  expect( getCritical( [
    'a'
  ] ) ).toMatchSnapshot()
} )

test( `preserve id`, () => {
  expect(
    compile( `#id { color: #111; }` )
  ).toMatchSnapshot()

  expect(
    compile( `.klass #id { color: #111; }` )
  ).toMatchSnapshot()

  expect(
    compile( `.klass, #id { color: #111; }` )
  ).toMatchSnapshot()
} )

test( `preserve tag`, () => {
  expect(
    compile( `tag { color: #111; }` )
  ).toMatchSnapshot()

  expect(
    compile( `.klass tag { color: #111; }` )
  ).toMatchSnapshot()

  expect(
    compile( `.klass, tag { color: #111; }` )
  ).toMatchSnapshot()
} )

test( `preserve universal`, () => {
  expect(
    compile( `* { color: #111; }` )
  ).toMatchSnapshot()

  expect(
    compile( `html * { color: #111; }` )
  ).toMatchSnapshot()

  expect(
    compile( `.klass * { color: #111; }` )
  ).toMatchSnapshot()

  expect(
    compile( `.klass, * { color: #111; }` )
  ).toMatchSnapshot()
} )

test( 'only class', () => {
  expect(
    compile( `.klass { color: #111; }` )
  ).toMatchSnapshot()

  expect(
    compile( `.klass::before { color: #111; }` )
  ).toMatchSnapshot()

  expect(
    compile( `.klass:hover { color: #111; }` )
  ).toMatchSnapshot()

  expect(
    compile( `.klass, .klass2 { color: #111; }` )
  ).toMatchSnapshot()

  expect(
    compile( `.klass > .klass2 { color: #111; }` )
  ).toMatchSnapshot()

  expect(
    compile( `.klass + .klass2 { color: #111; }` )
  ).toMatchSnapshot()

  expect(
    compile( `.klass ~ .klass2 { color: #111; }` )
  ).toMatchSnapshot()

  expect(
    compile( `.klass::before { color: #111; }` )
  ).toMatchSnapshot()
} )
