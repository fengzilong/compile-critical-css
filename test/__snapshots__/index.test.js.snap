// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`only class 1`] = `
"function getCriticalCSS( classlist = [] ) {
  const $in = ( classlist, selectorClasslist ) => {
    return classlist.some( klass => selectorClasslist.includes( klass ) )
  }
  return \`\${ $in( classlist, [\\"klass\\"] ) ? \\".klass { color: #111; }\\" : \\"\\" }\`
}"
`;

exports[`only class 2`] = `
"function getCriticalCSS( classlist = [] ) {
  const $in = ( classlist, selectorClasslist ) => {
    return classlist.some( klass => selectorClasslist.includes( klass ) )
  }
  return \`\${ $in( classlist, [\\"klass\\"] ) ? \\".klass::before { color: #111; }\\" : \\"\\" }\`
}"
`;

exports[`only class 3`] = `
"function getCriticalCSS( classlist = [] ) {
  const $in = ( classlist, selectorClasslist ) => {
    return classlist.some( klass => selectorClasslist.includes( klass ) )
  }
  return \`\${ $in( classlist, [\\"klass\\"] ) ? \\".klass:hover { color: #111; }\\" : \\"\\" }\`
}"
`;

exports[`only class 4`] = `
"function getCriticalCSS( classlist = [] ) {
  const $in = ( classlist, selectorClasslist ) => {
    return classlist.some( klass => selectorClasslist.includes( klass ) )
  }
  return \`\${ $in( classlist, [\\"klass\\",\\"klass2\\"] ) ? \\".klass, .klass2 { color: #111; }\\" : \\"\\" }\`
}"
`;

exports[`only class 5`] = `
"function getCriticalCSS( classlist = [] ) {
  const $in = ( classlist, selectorClasslist ) => {
    return classlist.some( klass => selectorClasslist.includes( klass ) )
  }
  return \`\${ $in( classlist, [\\"klass\\",\\"klass2\\"] ) ? \\".klass > .klass2 { color: #111; }\\" : \\"\\" }\`
}"
`;

exports[`only class 6`] = `
"function getCriticalCSS( classlist = [] ) {
  const $in = ( classlist, selectorClasslist ) => {
    return classlist.some( klass => selectorClasslist.includes( klass ) )
  }
  return \`\${ $in( classlist, [\\"klass\\",\\"klass2\\"] ) ? \\".klass + .klass2 { color: #111; }\\" : \\"\\" }\`
}"
`;

exports[`only class 7`] = `
"function getCriticalCSS( classlist = [] ) {
  const $in = ( classlist, selectorClasslist ) => {
    return classlist.some( klass => selectorClasslist.includes( klass ) )
  }
  return \`\${ $in( classlist, [\\"klass\\",\\"klass2\\"] ) ? \\".klass ~ .klass2 { color: #111; }\\" : \\"\\" }\`
}"
`;

exports[`only class 8`] = `
"function getCriticalCSS( classlist = [] ) {
  const $in = ( classlist, selectorClasslist ) => {
    return classlist.some( klass => selectorClasslist.includes( klass ) )
  }
  return \`\${ $in( classlist, [\\"klass\\"] ) ? \\".klass::before { color: #111; }\\" : \\"\\" }\`
}"
`;

exports[`preserve id 1`] = `
"function getCriticalCSS( classlist = [] ) {
  const $in = ( classlist, selectorClasslist ) => {
    return classlist.some( klass => selectorClasslist.includes( klass ) )
  }
  return \`#id { color: #111; }\`
}"
`;

exports[`preserve id 2`] = `
"function getCriticalCSS( classlist = [] ) {
  const $in = ( classlist, selectorClasslist ) => {
    return classlist.some( klass => selectorClasslist.includes( klass ) )
  }
  return \`.klass #id { color: #111; }\`
}"
`;

exports[`preserve id 3`] = `
"function getCriticalCSS( classlist = [] ) {
  const $in = ( classlist, selectorClasslist ) => {
    return classlist.some( klass => selectorClasslist.includes( klass ) )
  }
  return \`.klass, #id { color: #111; }\`
}"
`;

exports[`preserve tag 1`] = `
"function getCriticalCSS( classlist = [] ) {
  const $in = ( classlist, selectorClasslist ) => {
    return classlist.some( klass => selectorClasslist.includes( klass ) )
  }
  return \`tag { color: #111; }\`
}"
`;

exports[`preserve tag 2`] = `
"function getCriticalCSS( classlist = [] ) {
  const $in = ( classlist, selectorClasslist ) => {
    return classlist.some( klass => selectorClasslist.includes( klass ) )
  }
  return \`.klass tag { color: #111; }\`
}"
`;

exports[`preserve tag 3`] = `
"function getCriticalCSS( classlist = [] ) {
  const $in = ( classlist, selectorClasslist ) => {
    return classlist.some( klass => selectorClasslist.includes( klass ) )
  }
  return \`.klass, tag { color: #111; }\`
}"
`;

exports[`preserve universal 1`] = `
"function getCriticalCSS( classlist = [] ) {
  const $in = ( classlist, selectorClasslist ) => {
    return classlist.some( klass => selectorClasslist.includes( klass ) )
  }
  return \`* { color: #111; }\`
}"
`;

exports[`preserve universal 2`] = `
"function getCriticalCSS( classlist = [] ) {
  const $in = ( classlist, selectorClasslist ) => {
    return classlist.some( klass => selectorClasslist.includes( klass ) )
  }
  return \`html * { color: #111; }\`
}"
`;

exports[`preserve universal 3`] = `
"function getCriticalCSS( classlist = [] ) {
  const $in = ( classlist, selectorClasslist ) => {
    return classlist.some( klass => selectorClasslist.includes( klass ) )
  }
  return \`.klass * { color: #111; }\`
}"
`;

exports[`preserve universal 4`] = `
"function getCriticalCSS( classlist = [] ) {
  const $in = ( classlist, selectorClasslist ) => {
    return classlist.some( klass => selectorClasslist.includes( klass ) )
  }
  return \`.klass, * { color: #111; }\`
}"
`;
